File: v8-core/src/oracle/authority.rs
//! Counterfactual authority, identifiability, and fail-closed outcome semantics
//! (TARGET_ORACLE_SPEC §8, §16, §18.1).

#![allow(dead_code)]

use serde::{Deserialize, Serialize};

use crate::hash::Canon;

use super::artifacts::OracleEvaluationRecord;
use super::taxonomy::{AuthorityLevel, Identifiability, OracleContext, OracleRefusal, ValueNotion};

/// Explicit counterfactual authority composition. Authority level and
/// identifiability status are strictly orthogonal (TARGET_ORACLE_SPEC §8.3).
#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
pub struct CounterfactualAuthority {
    pub oracle_authority_level: AuthorityLevel,
    pub identifiability_status: Identifiability,
    pub support_rule_id: String,
    pub environment_model_id: String,
    pub assumptions: Vec<String>,
}

impl CounterfactualAuthority {
    pub fn new(
        oracle_authority_level: AuthorityLevel,
        identifiability_status: Identifiability,
        support_rule_id: impl Into<String>,
        environment_model_id: impl Into<String>,
        assumptions: Vec<String>,
    ) -> Self {
        Self {
            oracle_authority_level,
            identifiability_status,
            support_rule_id: support_rule_id.into(),
            environment_model_id: environment_model_id.into(),
            assumptions,
        }
    }

    pub fn identity(&self) -> String {
        let mut c = Canon::new();
        c.push_str("counterfactual-authority-v1");
        c.push_value(&serde_json::json!({
            "oracle_authority_level": format!("{:?}", self.oracle_authority_level),
            "identifiability_status": format!("{:?}", self.identifiability_status),
            "support_rule_id": self.support_rule_id,
            "environment_model_id": self.environment_model_id,
            "assumptions": self.assumptions,
        }));
        c.finish_sha1_hex()
    }

    pub fn is_identified(&self) -> bool {
        self.identifiability_status == Identifiability::Identified
    }
}

/// Typed Oracle evaluation outcome wrapping refusal and authority.
/// UNKNOWN is first-class and never collapses to a zero point estimate.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub enum OracleOutcome {
    Identified {
        point_estimate: f64,
        authority: CounterfactualAuthority,
    },
    PartiallyIdentified {
        lower_bound: f64,
        upper_bound: f64,
        authority: CounterfactualAuthority,
    },
    ModelDerived {
        point_estimate: Option<f64>,
        lower_bound: Option<f64>,
        upper_bound: Option<f64>,
        authority: CounterfactualAuthority,
    },
    Unknown {
        refusal: OracleRefusal,
        authority: CounterfactualAuthority,
    },
}

impl OracleOutcome {
    pub fn identified(
        point: f64,
        authority: CounterfactualAuthority,
    ) -> Result<Self, OracleRefusal> {
        if authority.identifiability_status != Identifiability::Identified {
            return Err(OracleRefusal::InsufficientSupport);
        }
        if !point.is_finite() {
            return Err(OracleRefusal::InsufficientSupport);
        }
        Ok(Self::Identified {
            point_estimate: point,
            authority,
        })
    }

    pub fn partially_identified(
        lower: f64,
        upper: f64,
        authority: CounterfactualAuthority,
    ) -> Result<Self, OracleRefusal> {
        if authority.identifiability_status != Identifiability::PartiallyIdentified {
            return Err(OracleRefusal::InsufficientSupport);
        }
        if !lower.is_finite() || !upper.is_finite() || lower > upper {
            return Err(OracleRefusal::InsufficientSupport);
        }
        Ok(Self::PartiallyIdentified {
            lower_bound: lower,
            upper_bound: upper,
            authority,
        })
    }

    pub fn model_derived(
        point: Option<f64>,
        lower: Option<f64>,
        upper: Option<f64>,
        authority: CounterfactualAuthority,
    ) -> Result<Self, OracleRefusal> {
        if authority.identifiability_status != Identifiability::ModelDerived {
            return Err(OracleRefusal::InsufficientSupport);
        }
        Ok(Self::ModelDerived {
            point_estimate: point,
            lower_bound: lower,
            upper_bound: upper,
            authority,
        })
    }

    pub fn unknown(refusal: OracleRefusal, authority: CounterfactualAuthority) -> Self {
        Self::Unknown {
            refusal,
            authority,
        }
    }

    pub fn is_unknown(&self) -> bool {
        matches!(self, Self::Unknown { .. })
    }

    pub fn is_identified(&self) -> bool {
        matches!(self, Self::Identified { .. })
    }

    pub fn point_estimate(&self) -> Option<f64> {
        match self {
            Self::Identified { point_estimate, .. } => Some(*point_estimate),
            Self::ModelDerived { point_estimate, .. } => *point_estimate,
            Self::PartiallyIdentified { .. } | Self::Unknown { .. } => None,
        }
    }

    pub fn bounds(&self) -> Option<(f64, f64)> {
        match self {
            Self::Identified { point_estimate, .. } => Some((*point_estimate, *point_estimate)),
            Self::PartiallyIdentified {
                lower_bound,
                upper_bound,
                ..
            } => Some((*lower_bound, *upper_bound)),
            Self::ModelDerived {
                lower_bound: Some(l),
                upper_bound: Some(u),
                ..
            } => Some((*l, *u)),
            _ => None,
        }
    }

    pub fn refusal_reason(&self) -> Option<OracleRefusal> {
        match self {
            Self::Unknown { refusal, .. } => Some(*refusal),
            _ => None,
        }
    }

    pub fn refusal_code(&self) -> Option<&'static str> {
        self.refusal_reason().map(|r| r.code())
    }

    pub fn authority(&self) -> &CounterfactualAuthority {
        match self {
            Self::Identified { authority, .. }
            | Self::PartiallyIdentified { authority, .. }
            | Self::ModelDerived { authority, .. }
            | Self::Unknown { authority, .. } => authority,
        }
    }

    /// Compare outcomes for decision ordering.
    /// Overlapping intervals fail closed with `OracleRefusal::InsufficientSupport`.
    /// UNKNOWN and ModelDerived fail closed with their respective refusal.
    pub fn compare_for_ordering(&self, other: &Self) -> Result<std::cmp::Ordering, OracleRefusal> {
        match (self, other) {
            (Self::Unknown { refusal, .. }, _) => Err(*refusal),
            (_, Self::Unknown { refusal, .. }) => Err(*refusal),
            (Self::ModelDerived { .. }, _) | (_, Self::ModelDerived { .. }) => {
                Err(OracleRefusal::ModelOnlyCounterfactual)
            }
            (
                Self::Identified {
                    point_estimate: a, ..
                },
                Self::Identified {
                    point_estimate: b, ..
                },
            ) => a
                .partial_cmp(b)
                .ok_or(OracleRefusal::InsufficientSupport),
            (a, b) => {
                let (a_lo, a_hi) = a.bounds().ok_or(OracleRefusal::InsufficientSupport)?;
                let (b_lo, b_hi) = b.bounds().ok_or(OracleRefusal::InsufficientSupport)?;
                if a_hi < b_lo {
                    Ok(std::cmp::Ordering::Less)
                } else if a_lo > b_hi {
                    Ok(std::cmp::Ordering::Greater)
                } else {
                    // Overlapping intervals fail closed
                    Err(OracleRefusal::InsufficientSupport)
                }
            }
        }
    }

    pub fn to_evaluation_record(
        &self,
        context: &OracleContext,
        candidate_population_hash: &str,
        action_manifest_hash: &str,
        simulator_or_receipt_hash: &str,
        code_hash: &str,
        config_hash: &str,
        value_notion: ValueNotion,
        lineage_id: &str,
    ) -> OracleEvaluationRecord {
        let auth = self.authority();
        let (pt, lo, hi) = match self {
            Self::Identified { point_estimate, .. } => (Some(*point_estimate), None, None),
            Self::PartiallyIdentified {
                lower_bound,
                upper_bound,
                ..
            } => (None, Some(*lower_bound), Some(*upper_bound)),
            Self::ModelDerived {
                point_estimate,
                lower_bound,
                upper_bound,
                ..
            } => (*point_estimate, *lower_bound, *upper_bound),
            Self::Unknown { .. } => (None, None, None),
        };

        let mut rec = OracleEvaluationRecord {
            evaluation_id: String::new(),
            oracle_role: context.role,
            authority_level: auth.oracle_authority_level,
            identifiability_status: auth.identifiability_status,
            information_contract_id: context.information_contract_id.clone(),
            opportunity_universe_id: context.opportunity_universe_id.clone(),
            utility_contract_id: context.utility_contract_id.clone(),
            policy_class_id: context.policy_class_id.clone(),
            cost_model_id: context.cost_model_id.clone(),
            capacity_model_id: context.capacity_model_id.clone(),
            environment_target_id: context.environment_target_id.clone(),
            candidate_population_hash: candidate_population_hash.to_string(),
            action_manifest_hash: action_manifest_hash.to_string(),
            simulator_or_receipt_hash: simulator_or_receipt_hash.to_string(),
            code_hash: code_hash.to_string(),
            config_hash: config_hash.to_string(),
            value_notion,
            point_estimate: pt,
            lower_bound: lo,
            upper_bound: hi,
            uncertainty_artifact_id: None,
            refusal_reason: self.refusal_code().map(ToString::to_string),
            assumptions: auth.assumptions.clone(),
            lineage_id: lineage_id.to_string(),
        };
        rec.bind_identity();
        rec
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn sample_authority(level: AuthorityLevel, ident: Identifiability) -> CounterfactualAuthority {
        CounterfactualAuthority::new(
            level,
            ident,
            "l1-support-v1",
            "env-model-v1",
            vec!["L1_BAR_CLOSE".into()],
        )
    }

    #[test]
    fn unknown_returns_no_numeric_point_value() {
        let auth = sample_authority(AuthorityLevel::L1, Identifiability::NotIdentifiable);
        let outcome = OracleOutcome::unknown(OracleRefusal::MissingDecisionTimeData, auth);
        assert!(outcome.is_unknown());
        assert_eq!(outcome.point_estimate(), None);
        assert_eq!(outcome.bounds(), None);
        assert_eq!(
            outcome.refusal_code(),
            Some("MISSING_DECISION_TIME_DATA")
        );
    }

    #[test]
    fn identified_outcome_carries_point_estimate() {
        let auth = sample_authority(AuthorityLevel::L1, Identifiability::Identified);
        let outcome = OracleOutcome::identified(1.25, auth).unwrap();
        assert!(outcome.is_identified());
        assert_eq!(outcome.point_estimate(), Some(1.25));
        assert_eq!(outcome.bounds(), Some((1.25, 1.25)));
        assert_eq!(outcome.refusal_reason(), None);
    }

    #[test]
    fn partially_identified_overlapping_intervals_fail_closed_for_ordering() {
        let auth = sample_authority(AuthorityLevel::L1, Identifiability::PartiallyIdentified);
        let a = OracleOutcome::partially_identified(0.5, 1.5, auth.clone()).unwrap();
        let b = OracleOutcome::partially_identified(1.0, 2.0, auth.clone()).unwrap();
        let c = OracleOutcome::partially_identified(2.5, 3.5, auth).unwrap();

        // Disjoint intervals compare successfully
        assert_eq!(a.compare_for_ordering(&c), Ok(std::cmp::Ordering::Less));
        assert_eq!(c.compare_for_ordering(&a), Ok(std::cmp::Ordering::Greater));

        // Overlapping intervals fail closed
        assert_eq!(
            a.compare_for_ordering(&b),
            Err(OracleRefusal::InsufficientSupport)
        );
    }

    #[test]
    fn unknown_fails_closed_in_ordering_comparison() {
        let auth_id = sample_authority(AuthorityLevel::L1, Identifiability::Identified);
        let auth_unk = sample_authority(AuthorityLevel::L1, Identifiability::NotIdentifiable);
        let good = OracleOutcome::identified(1.0, auth_id).unwrap();
        let bad = OracleOutcome::unknown(OracleRefusal::ExecutionAuthorityTooWeak, auth_unk);

        assert_eq!(
            good.compare_for_ordering(&bad),
            Err(OracleRefusal::ExecutionAuthorityTooWeak)
        );
        assert_eq!(
            bad.compare_for_ordering(&good),
            Err(OracleRefusal::ExecutionAuthorityTooWeak)
        );
    }
}

File: v8-next/src/v8_next/oracle/authority.py
"""Counterfactual authority and fail-closed outcome semantics — port of
v8-core/src/oracle/authority.rs (TARGET_ORACLE_SPEC §8, §16, §18.1).

Authority level and identifiability status are strictly orthogonal. ``OracleOutcome`` is a
closed set of four shapes and **UNKNOWN is first-class**: it carries a typed refusal and never
collapses into a zero point estimate. Every constructor refuses the states the Rust side
refuses, with the same refusal code (``INSUFFICIENT_SUPPORT``).
"""

from __future__ import annotations

import hashlib
import json
import math
from dataclasses import dataclass, field
from enum import StrEnum

from v8_next.oracle.artifacts import OracleEvaluationRecord
from v8_next.oracle.taxonomy import (
    AuthorityLevel,
    Identifiability,
    OracleContext,
    OracleRefusal,
    ValueNotion,
)


class OracleRefused(Exception):
    """Raised when an outcome is asked for a state the data does not support."""

    def __init__(self, refusal: OracleRefusal, message: str = "") -> None:
        super().__init__(f"{refusal.value}{': ' + message if message else ''}")
        self.refusal = refusal


@dataclass(frozen=True)
class CounterfactualAuthority:
    """Explicit counterfactual authority composition (the Rust *struct*, not the axis enum)."""

    oracle_authority_level: AuthorityLevel
    identifiability_status: Identifiability
    support_rule_id: str
    environment_model_id: str
    assumptions: tuple[str, ...] = ()

    def identity(self) -> str:
        payload = json.dumps(
            {
                "oracle_authority_level": self.oracle_authority_level.name,
                "identifiability_status": self.identifiability_status.value,
                "support_rule_id": self.support_rule_id,
                "environment_model_id": self.environment_model_id,
                "assumptions": list(self.assumptions),
            },
            sort_keys=True,
            separators=(",", ":"),
        )
        return hashlib.sha1(f"counterfactual-authority-v1|{payload}".encode()).hexdigest()

    def is_identified(self) -> bool:
        return self.identifiability_status is Identifiability.IDENTIFIED

    def as_dict(self) -> dict[str, object]:
        return {
            "oracle_authority_level": self.oracle_authority_level.name,
            "identifiability_status": self.identifiability_status.value,
            "support_rule_id": self.support_rule_id,
            "environment_model_id": self.environment_model_id,
            "assumptions": list(self.assumptions),
            "identity": self.identity(),
        }


class OracleOutcomeKind(StrEnum):
    IDENTIFIED = "IDENTIFIED"
    PARTIALLY_IDENTIFIED = "PARTIALLY_IDENTIFIED"
    MODEL_DERIVED = "MODEL_DERIVED"
    UNKNOWN = "UNKNOWN"


@dataclass(frozen=True)
class OracleOutcome:
    """A counterfactual answer, or a typed refusal. UNKNOWN is never a zero."""

    kind: OracleOutcomeKind
    authority: CounterfactualAuthority
    point_estimate: float | None = None
    lower_bound: float | None = None
    upper_bound: float | None = None
    refusal: OracleRefusal | None = None
    supporting_evidence: dict[str, object] = field(default_factory=dict)

    # -- constructors, each refusing what the Rust side refuses ----------------------

    @classmethod
    def identified(cls, point: float, authority: CounterfactualAuthority) -> OracleOutcome:
        if authority.identifiability_status is not Identifiability.IDENTIFIED:
            raise OracleRefused(OracleRefusal.INSUFFICIENT_SUPPORT, "authority is not identified")
        if not math.isfinite(point):
            raise OracleRefused(OracleRefusal.INSUFFICIENT_SUPPORT, "non-finite point estimate")
        return cls(OracleOutcomeKind.IDENTIFIED, authority, point_estimate=point)

    @classmethod
    def partially_identified(
        cls, lower: float, upper: float, authority: CounterfactualAuthority
    ) -> OracleOutcome:
        if authority.identifiability_status is not Identifiability.PARTIALLY_IDENTIFIED:
            raise OracleRefused(OracleRefusal.INSUFFICIENT_SUPPORT, "authority is not partial")
        if not (math.isfinite(lower) and math.isfinite(upper)) or lower > upper:
            raise OracleRefused(OracleRefusal.INSUFFICIENT_SUPPORT, "invalid bounds")
        return cls(
            OracleOutcomeKind.PARTIALLY_IDENTIFIED, authority, lower_bound=lower, upper_bound=upper
        )

    @classmethod
    def model_derived(
        cls,
        point: float | None,
        lower: float | None,
        upper: float | None,
        authority: CounterfactualAuthority,
    ) -> OracleOutcome:
        if authority.identifiability_status is not Identifiability.MODEL_DERIVED:
            raise OracleRefused(OracleRefusal.INSUFFICIENT_SUPPORT, "authority is not model-derived")
        return cls(
            OracleOutcomeKind.MODEL_DERIVED,
            authority,
            point_estimate=point,
            lower_bound=lower,
            upper_bound=upper,
        )

    @classmethod
    def unknown(
        cls, refusal: OracleRefusal, authority: CounterfactualAuthority
    ) -> OracleOutcome:
        return cls(OracleOutcomeKind.UNKNOWN, authority, refusal=refusal)

    # -- read side --------------------------------------------------------------------

    def is_unknown(self) -> bool:
        return self.kind is OracleOutcomeKind.UNKNOWN

    def is_identified(self) -> bool:
        return self.kind is OracleOutcomeKind.IDENTIFIED

    def bounds(self) -> tuple[float, float] | None:
        if self.lower_bound is None or self.upper_bound is None:
            return None
        return (self.lower_bound, self.upper_bound)

    def refusal_reason(self) -> OracleRefusal | None:
        return self.refusal

    def to_evaluation_record(
        self,
        context: OracleContext,
        candidate_population_hash: str,
        action_manifest_hash: str,
        simulator_or_receipt_hash: str,
        code_hash: str,
        config_hash: str,
        value_notion: ValueNotion,
        lineage_id: str,
    ) -> OracleEvaluationRecord:
        """Bind this outcome to its authority lineage (port of the Rust method).

        UNKNOWN carries no point estimate and no bounds — the refusal is the record.
        """
        if self.kind is OracleOutcomeKind.IDENTIFIED:
            point, lower, upper = self.point_estimate, None, None
        elif self.kind is OracleOutcomeKind.PARTIALLY_IDENTIFIED:
            point, lower, upper = None, self.lower_bound, self.upper_bound
        elif self.kind is OracleOutcomeKind.MODEL_DERIVED:
            point, lower, upper = self.point_estimate, self.lower_bound, self.upper_bound
        else:
            point, lower, upper = None, None, None
        record = OracleEvaluationRecord(
            oracle_role=context.role,
            authority_level=self.authority.oracle_authority_level,
            identifiability_status=self.authority.identifiability_status,
            information_contract_id=context.information_contract_id,
            opportunity_universe_id=context.opportunity_universe_id,
            utility_contract_id=context.utility_contract_id,
            policy_class_id=context.policy_class_id,
            cost_model_id=context.cost_model_id,
            capacity_model_id=context.capacity_model_id,
            environment_target_id=context.environment_target_id,
            candidate_population_hash=candidate_population_hash,
            action_manifest_hash=action_manifest_hash,
            simulator_or_receipt_hash=simulator_or_receipt_hash,
            code_hash=code_hash,
            config_hash=config_hash,
            value_notion=value_notion,
            point_estimate=point,
            lower_bound=lower,
            upper_bound=upper,
            refusal_reason=None if self.refusal is None else self.refusal.value,
            assumptions=list(self.authority.assumptions),
            lineage_id=lineage_id,
        )
        record.bind_identity()
        return record

    def as_dict(self) -> dict[str, object]:
        return {
            "kind": self.kind.value,
            "point_estimate": self.point_estimate,
            "bounds": None if self.bounds() is None else list(self.bounds() or ()),
            "refusal": None if self.refusal is None else self.refusal.value,
            "authority": self.authority.as_dict(),
            "supporting_evidence": self.supporting_evidence,
            "note": (
                "an UNKNOWN outcome carries a typed refusal; it is never a zero point estimate"
                if self.is_unknown()
                else None
            ),
        }
